File: android/modules/open-brewery-sdk/src/main/java/com/multiplatformbusinesssdk/openbrewery/OpenBrewerySDK.kt
package com.multiplatformbusinesssdk.openbrewery

import com.multiplatformbusinesssdk.core.ApiRequest
import com.multiplatformbusinesssdk.core.HttpMethod
import com.multiplatformbusinesssdk.core.NetworkEngine

object OpenBrewerySDKConfig {
    const val baseUrl: String = "https://api.openbrewerydb.org/v1"
}

data class BrewerySummary(
    val id: String,
    val name: String,
    val brewery_type: String,
    val city: String,
    val state: String,
    val country: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): BrewerySummary {
                        val id = json.getString("id")
            val name = json.getString("name")
            val brewery_type = json.getString("brewery_type")
            val city = json.getString("city")
            val state = json.getString("state")
            val country = json.getString("country")
            return BrewerySummary(id, name, brewery_type, city, state, country)
        }
    }
}

data class BreweryDetailResponse(
    val id: String,
    val name: String,
    val brewery_type: String,
    val address_1: String,
    val city: String,
    val state_province: String,
    val postal_code: String,
    val country: String,
    val phone: String,
    val website_url: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): BreweryDetailResponse {
                        val id = json.getString("id")
            val name = json.getString("name")
            val brewery_type = json.getString("brewery_type")
            val address_1 = json.getString("address_1")
            val city = json.getString("city")
            val state_province = json.getString("state_province")
            val postal_code = json.getString("postal_code")
            val country = json.getString("country")
            val phone = json.getString("phone")
            val website_url = json.getString("website_url")
            return BreweryDetailResponse(id, name, brewery_type, address_1, city, state_province, postal_code, country, phone, website_url)
        }
    }
}


private class BreweryRepositoryFetchBreweriesTestRequest(private val by_city: String, private val per_page: Int = 10) : ApiRequest<List<BrewerySummary>> {
    override val path: String = "/breweries"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = mapOf(
            "by_city" to by_city,
            "per_page" to per_page.toString()
        )
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): List<BrewerySummary> =
        run {
    val items = org.json.JSONArray(payload).let { payload ->
        when (payload) {
            is org.json.JSONArray -> payload
            is org.json.JSONObject -> payload.getJSONArray("data")
            else -> throw IllegalArgumentException("Unsupported JSON payload")
        }
    }
    buildList {
        for (index in 0 until items.length()) {
            add(BrewerySummary.fromJson(items.getJSONObject(index)))
        }
    }
}
}

private class BreweryRepositoryFetchBreweryDetailRequest(private val id: String) : ApiRequest<BreweryDetailResponse> {
    override val path: String = "/breweries/$id"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): BreweryDetailResponse =
        BreweryDetailResponse.fromJson(org.json.JSONObject(payload) as org.json.JSONObject)
}


interface BreweryRepositoryProtocol {
    suspend fun fetchBreweriesTest(by_city: String, per_page: Int): List<BrewerySummary>
    suspend fun fetchBreweryDetail(id: String): BreweryDetailResponse
}

class BreweryRepository(
    private val networkEngine: NetworkEngine
) : BreweryRepositoryProtocol {
override suspend fun fetchBreweriesTest(by_city: String, per_page: Int): List<BrewerySummary> {
    return networkEngine.request(BreweryRepositoryFetchBreweriesTestRequest(by_city = by_city, per_page = per_page))
}

override suspend fun fetchBreweryDetail(id: String): BreweryDetailResponse {
    return networkEngine.request(BreweryRepositoryFetchBreweryDetailRequest(id = id))
}

}


interface OpenBrewerySDKClientProtocol {
    val breweryRepository: BreweryRepositoryProtocol
}

class OpenBrewerySDKClient(
    networkEngine: NetworkEngine = NetworkEngine(baseUrl = OpenBrewerySDKConfig.baseUrl)
) : OpenBrewerySDKClientProtocol {
    override val breweryRepository: BreweryRepositoryProtocol = BreweryRepository(networkEngine)
}

File: BusinessSDK/Sources/OpenBrewerySDK/OpenBrewerySDK.swift
import Foundation
import BusinessSDKCore

public enum OpenBrewerySDKConfig {
    public static let baseURL = URL(string: "https://api.openbrewerydb.org/v1")!
}

public struct BrewerySummary: Codable, Sendable {
    public let id: String
    public let name: String
    public let brewery_type: String
    public let city: String
    public let state: String
    public let country: String

    public init(id: String, name: String, brewery_type: String, city: String, state: String, country: String) {
        self.id = id
        self.name = name
        self.brewery_type = brewery_type
        self.city = city
        self.state = state
        self.country = country
    }
}

public struct BreweryDetailResponse: Codable, Sendable {
    public let id: String
    public let name: String
    public let brewery_type: String
    public let address_1: String
    public let city: String
    public let state_province: String
    public let postal_code: String
    public let country: String
    public let phone: String
    public let website_url: String

    public init(id: String, name: String, brewery_type: String, address_1: String, city: String, state_province: String, postal_code: String, country: String, phone: String, website_url: String) {
        self.id = id
        self.name = name
        self.brewery_type = brewery_type
        self.address_1 = address_1
        self.city = city
        self.state_province = state_province
        self.postal_code = postal_code
        self.country = country
        self.phone = phone
        self.website_url = website_url
    }
}


private struct BreweryRepositoryFetchBreweriesTestRequest: APIRequest {
    private let by_city: String
    private let per_page: Int

    init(by_city: String, per_page: Int = 10) {
        self.by_city = by_city
        self.per_page = per_page
    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/breweries"
    }

    var queryItems: [URLQueryItem] {
        [URLQueryItem(name: "by_city", value: String(describing: by_city)), URLQueryItem(name: "per_page", value: String(describing: per_page))]
    }

    typealias Response = [BrewerySummary]
}

private struct BreweryRepositoryFetchBreweryDetailRequest: APIRequest {
    private let id: String

    init(id: String) {
        self.id = id
    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/breweries/\(id)"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = BreweryDetailResponse
}


public protocol BreweryRepositoryProtocol: Sendable {
    func fetchBreweriesTest(by_city: String, per_page: Int) async throws -> [BrewerySummary]
    func fetchBreweryDetail(id: String) async throws -> BreweryDetailResponse
}

public final class BreweryRepository: BreweryRepositoryProtocol, @unchecked Sendable {
    private let networkEngine: NetworkEngineProtocol

    public init(networkEngine: NetworkEngineProtocol) {
        self.networkEngine = networkEngine
    }

public func fetchBreweriesTest(by_city: String, per_page: Int = 10) async throws -> [BrewerySummary] {
    try await networkEngine.request(BreweryRepositoryFetchBreweriesTestRequest(by_city: by_city, per_page: per_page))
}

public func fetchBreweryDetail(id: String) async throws -> BreweryDetailResponse {
    try await networkEngine.request(BreweryRepositoryFetchBreweryDetailRequest(id: id))
}

}


public protocol OpenBrewerySDKClientProtocol: Sendable {
    var breweryRepository: BreweryRepositoryProtocol { get }
}

public struct OpenBrewerySDKClient: OpenBrewerySDKClientProtocol, Sendable {
    public let breweryRepository: BreweryRepositoryProtocol

    public init(networkEngine: NetworkEngineProtocol = NetworkEngine(baseURL: OpenBrewerySDKConfig.baseURL)) {
        self.breweryRepository = BreweryRepository(networkEngine: networkEngine)
    }
}
